Feature/#6 planner api 개발

GahBaek opened the pull request:
## Description
planner api 관련 파일 생성

## Changes
controller
- [x] PlannerController 생성 

service
- [x] PlannerService 생성 

repository
- [x] PlannerRepository 생성

dto
- [x] PlannerDto
- [x] PlannerResponsedata
- [x] CompletedPlannerResponsedata

## Additional context
Closed #6

Commit:
feature: planner api 개발 - #6

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/repository/PlannerRepository.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.UUID;
+
+public interface PlannerRepository extends JpaRepository<Planner, UUID> {
+    Planner findPlannerById(UUID id);
```
JPA가 기본으로 제공하는 메서드 중에 findById()있어서 이건 없어도 될 것 같습니당

GahBaek replied:
기본 제공 되는 findById는 optional 반환이어서 planner을 반환하는 메서드를 사용하고 싶었습니당

GahBaek replied:
optional를 잘 사용하지 않았었거든용...

rnqhstmd replied:
optional 사용해서 findById(plannerId).orElseThrow(new NotFoundException) 이런식으로 예외처리 용이하게 하면 되지 않을까요??

GahBaek replied:
넵 !

rnqhstmd replied:
그리고 커밋할 때 좌측 탭에 commit 들어가면 파일 하나씩 커밋 찍을 수 있고 커밋 메세지 적을 때 엔터 눌러서 밑에 커밋 상세 메세지 적을 수 있는 거 기억하시죠..?ㅠㅅㅠ 지금 한 번에 다 커밋 들어간 것 같아요

GahBaek replied:
아 넵 지금 확인했습니다 !

rnqhstmd reviewed (commented):
확인 한 번 해주세요!!

Commit:
Merge branch 'main' of https://github.com/mju-likelion/12th-hackathon-team2-server into feature/planner-api-#6

Commit:
feature: planner controller 수정 - #6

Commit:
feature: planner api 개발 - #6

GahBaek commented:
아직 수정 중입니다 --- !!

Commit:
feature: planner api 유저 annotation 추가 - #6

PlannerController AuthenticatedUser 추가

Commit:
feature: PlannerService 에 AuthenticatedUser 적용

runtimeexception 으로 접근 권한 제한
본인의 플래너에 대한 수정, 조회, 완료 상태 변경, 완료된 플랜 조회 가능.

Commit:
feature: 잘못된 plannerId 추가 - #6

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/controller/PlannerController.java:
```diff
@@ -0,0 +1,60 @@
+
+    // 새 planner 생성
+    @PostMapping
+    public ResponseEntity<ResponseDto<Void>> createPlanner(@AuthenticatedUser User user, @RequestBody PlannerDto plannerDto) {
```
Dto 앞에 Valid 어노테이션 빠져있습니다 다른 api들 파라미터에도 추가해주세욥

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,107 @@
+        for (Planner planner : allPlanners) {
+            if(!planner.isCompleted())
+            {
+                PlannerResponseData plannerResponseData = PlannerResponseData.builder()
```
이런 생성 로직은 서비스계층이 아닌 PlannerResponseData클래스에 정적메서드 안에다 넣어주시면 좋습니다!

rnqhstmd replied:
근데 여태까지 저도 생성자 안에다 로직 넣어놨는데 구글링 해보니까 생성자방식보다 정적 팩토리 메서드 방식이 이점이 더 많더라구요..! 메서드 이름도 지정 가능, 파라미터도 다르게 해서 반환타입도 다르게할 수 있고, 호출할 때마다 new 로 객체 생성 안 해도 돼서 리소스 사용도 줄이고요

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public Planner updatePlanner(PlannerDto plannerDto, UUID plannerId, User user) {
+        Optional<Planner> optionalPlanner = plannerRepository.findById(plannerId);
```
plannerRepository.findById(plannerId).orElseThrow(()->NotFoundException(ErrorCode.PLANNER_NOT_FOUND))
이런식으로 한 줄로 줄일 수 있습니다

rnqhstmd replied:
차라리 이걸 스태틱 메서드 빼서 반환타입에 Planner 로 해주시고 Planner planner = 메서드이름(plannerId); 이런식으로 하면 코드를 더 줄일 수 있을 것 같아요 다른 기능에서도 이 메서드 쓸 일이 있으니까

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,107 @@
+
+    public Planner completePlannerById(UUID plannerId, User user) {
+        Optional<Planner> optionalPlanner = plannerRepository.findById(plannerId);
+        if (optionalPlanner.isPresent() && optionalPlanner.get().getUser().equals(user)) {
```
객체 전체를 비교하는 것보다 userId 필드만 비교하는 게 성능상 유리할 것 같습니다

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,107 @@
+        List<CompletedPlannerResponse> completedPlannerList = new ArrayList<>();
+        List<Planner> completedPlanners = plannerRepository.findByIsCompleted(true);
+        for (Planner planner : completedPlanners) {
+            CompletedPlannerResponse plannerResponseData = CompletedPlannerResponse.builder()
```
이것도 마찬가지로 클래스 내 생성자 안에 로직 넣어주시면 될 것 같습니다

rnqhstmd reviewed (commented):
이렇게 다시 확인 부탁드립니다!

GahBaek commented:
넵 지금 바로 수정하겠습니당 !

Commit:
feature: dto 생성자 추가 - #6

Commit:
feature: merge - #6
Merge branch 'main' of https://github.com/mju-likelion/12th-hackathon-team2-server into feature/planner-api-#6

Commit:
feature: completedPlannerListData 클래스 추가 - #6

Commit:
feature: completedPlanner user 수정 - #6

Commit:
feature: getAllCompletedPlanners 메소드 수정 - #6

completedResponseListData class 삭제 및 생성자 수정

Commit:
feature: updatePlanner 메서드 수정 - #6

Commit:
feature: 정렬 수정 - #6

Commit:
feature: exception 추가 - #6

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+    public Planner updatePlanner(PlannerDto plannerDto, UUID plannerId, User user) {
+        Planner planner = findPlanner(plannerId);
+
+        if (!planner.getUser().getEmail().equals(user.getEmail())) {
```
email 필드로 작성자 식별하는 것보다 PK 값인 userId값으로 하는 게 낫지않을까요??

rnqhstmd replied:
그리고 다른 메서드에서도 반복되는 로직이면 스태틱 메서드로 만들어서 써야합니다!

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+        return plannerRepository.save(planner);
+    }
+
+    public Map<LocalDate, List<CompletedPlannerResponse>> getCompletedPlannersGroupedByDate(User user) {
```
이건 날짜별로 해당 날짜에 완료된 플래너를 묶어서 보여주는 건가요??

rnqhstmd replied:
작동 되나요??

GahBaek replied:
다른 날짜에 저장 된 건 아직 테스트를 안해봤는데 오늘 7.18 저장 한다면 7.18로 저장 됩니다. notion에 output 저장해 놨어요

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+                .collect(Collectors.groupingBy(planner -> planner.getModifiedDate().toLocalDate()));
+    }
+
+    private List<CompletedPlannerResponse> getAllCompletedPlanners(User user) {
```
plannerResponse를 묶어서 List로 만드는 Dto를 따로 만들어서 생성자에 로직 넣는 건 어떨까유

rnqhstmd replied:
지금 보니까 서비스 계층에서 Response들을 List로 반환해야하는 메서드들을 ArrayList로 만들어 바로 반환하고있는데 List로 바꿔주는 별도의 Dto를 만들어 컨트롤러에서 쓸 수 있게 반환해줘야해요

rnqhstmd replied:
여태까지 저도 생성자 안에다 로직 넣어놨는데 구글링 해보니까 생성자방식보다 정적 팩토리 메서드 방식이 이점이 더 많더라구요..! 메서드 이름도 지정 가능, 파라미터도 다르게 해서 반환타입도 다르게할 수 있고, 호출할 때마다 new 로 객체 생성 안 해도 돼서 리소스 사용도 줄이고요

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+    public Map<LocalDate, List<CompletedPlannerResponse>> getCompletedPlannersGroupedByDate(User user) {
+        List<CompletedPlannerResponse> completedPlannerList = getAllCompletedPlanners(user);
+
+        return completedPlannerList.stream()
```
음 이렇게 되면 컨트롤러에서 Map<LocalDate, List<CompletedPlannerResponse>>로 받게되니까 dto로 감싸서 보내는 게 낫지않을까요?

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        planner.setCompleted(true);
+        return plannerRepository.save(planner);
```
이렇게 되면 Planner 객체를 반환하게되는데 컨트롤러 플래너 완료 api 메서드에서 반환한 Planner 객체 안쓰이고 있잖아요 그럼 저장만 하고 반환타입을 void로 바꿔야할 것 같아요

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,92 @@
+                plannerList.add(plannerResponseData);
+            }
+        }
+        return plannerList;
```
List<PlannerResponseData> 받아 저장하는 PlannerResponseDataList라는 dto 만들어서 반환하는 게 낫지 않을까요? 어찌됐건 컨트롤러로 전달해야하는 거니까

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/controller/PlannerController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    // planner 완료하기
+    @PutMapping("/{plannerId}/complete")
```
완료하기도 수정하기랑 비슷한 기능이니 /{plannerId} 이렇게 바꾸는게 좋을 것 같습니다

rnqhstmd replied:
그럼 이것도 클라이언트에게서 받는 requestDto가 따로 필요할 것 같네요
체크박스(완료버튼) 누르면
{
  "isCompleted":"true"
}
머 이런식으로 응답받으면 완료하기 로직 들가게끔요

rnqhstmd reviewed (commented):
확인 부탁드립니다!

Commit:
feature: completePlanner 수정 - #6

Commit:
feature: complete, dto 추가 - #6

PlannerResponseListData 추가
completePlanner 수정

Commit:
feature: dto 추가 - #6
GroupedCompletedPlannerResponse, PlannerListResponseData, CompletePlannerRequestDto 추가

Commit:
feature: dto, responsedata 사용으로 Service 수정 - #6

Commit:
feature: user 비교 방법 수정 - #6

Commit:
feature: 정적 팩토리 메서드 방식 사용 - #6
CompletedPlannerResponse, PlannerResponseData 클래스에 적용

Commit:
feature: 정적 팩토리 메서드 방식 - #6

Commit:
feature: completePlanner requestDto 수정 - #6
boolean -> Boolean 변경

Commit:
feature: Planner 상태 변경 관련 수정 - #6
상태 변경 시 요청 상태가 false 인경우 혹은 null인경우 예외 발생 추가
이에 맞는 errorCode 추가

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,96 @@
+        this.plannerRepository.save(planner);
+    }
+
+    public Planner updatePlanner(PlannerDto plannerDto, UUID plannerId, User user) {
```
반환 타입을 void로 바꿔야할 것 같습니다

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,96 @@
+            throw new UnauthorizedException(ErrorCode.INVALID_PLANNER_ACCESS);
+        }
+        Optional<Planner> optionalPlanner = plannerRepository.findById(plannerId);
+        if (optionalPlanner.isPresent()) {
```
Planner 객체 뽑는 코드를 plannerRepo.findById(id).orElseThrow(()->new Exception())으로 바꿔보시고 option cmd v 누르면 변수가 추출됩니다. 한 번 해보시면 제가 여태까지 말했던 게 한 방에 이해갈 거에요! 해보시면 알겠지만 바로 Planner 객체가 반환됩니다 옵셔널 없이요 그럼 isPresent, get()과 같은 작업도 필요없어지는 겁니다

rnqhstmd replied:
밑에 findPlanner라는 메서드를 잘 만들어놓고 여기선 왜 활용 안 하신거에요ㅠㅠ

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,96 @@
+            planner.setCompleted(requestDto.getIsCompleted());
+            plannerRepository.save(planner);
+        } else {
+            throw new RuntimeException("Planner not found");
```
여기 else에서 던지는 not found 예외도 위 리뷰에서 말씀드린 코드 한 줄로 해결되겠죠??

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,96 @@
+            throw new UnauthorizedException(ErrorCode.NO_ACCESS, "해당 플래너에 접근 할 수 없습니다.");
+        }
+
+        planner.setContent(plannerDto.getContent());
```
플래너를 수정하는 건 완료 전 플랜일테니까 isCompleted가 false인지 먼저 확인하고 수정 진행해야하지 않을까여

rnqhstmd commented on src/main/java/com/example/mutsideout_mju/service/PlannerService.java:
```diff
@@ -0,0 +1,96 @@
+
+    private List<CompletedPlannerResponse> getAllCompletedPlanners(User user) {
+        List<Planner> completedPlanners = plannerRepository.findByIsCompletedAndUser(true, user);
+        return completedPlanners.stream()
```
여기도 getAllPlanners() 메서드처럼 CompletedPlannerResponse들을 List로 받는 response 클래스 하나 만들어서 반환하게끔해주세요

rnqhstmd reviewed (commented):
확인 부탁드립니당

Commit:
feature: 예외 수정 - #6

Commit:
feature: dto 추가

Commit:
feature: dto 수정 - #6
CompletedPlannerListResponse 추가 및 관련 메소드 수정

Commit:
feature: 사용자 확인 추가 - #6
사용자 확인 중복 메서드 추가

Commit:
feature: 상태 변경 시 예외 수정 - #6

GahBaek commented:
수정 했습니다 !

aaahyunseo commented:
저는 확인 완료했습니다!
+ commit 메세지 제목은 feat: 으로 작성해주시면 좋을거 같아요!
